src/core/lib/json/json_object_loader.h

```diff
@@ -0,0 +1,387 @@
+// Copyright 2020 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_JSON_JSON_OBJECT_LOADER_H
+#define GRPC_CORE_LIB_JSON_JSON_OBJECT_LOADER_H
+
+#include <grpc/support/port_platform.h>
+
+#include <sys/types.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <string>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/json/json.h"
+
+// Provides a means to load JSON objects into C++ objects, with the aim of
+// minimizing object code size.
+//
+// Usage:
+// Given struct Foo:
+//   struct Foo {
+//     int a;
+//     int b;
+//   };
+// We add a member to Foo to declare how to load the object from JSON:
+//   struct Foo {
+//     int a;
+//     int b;
+//     static const JsonLoaderInterface* JsonLoader() {
+//       static const auto loader = JsonObjectLoader<Foo>()
+//           .Field("a", &Foo::a)
+//           .Field("b", &Foo::b)
+//           .Finish();
```

Comment:
I think we may need some hooks for customizing validation for one-off use-cases.  One use-case for this is where we have a validation that needs to look at multiple fields.  For example, there's a case in the RLS LB policy where we clamp one field to the value of another field:

https://github.com/grpc/grpc/blob/28db143b98682b06aa5c58059ffe53b42f28b395/src/core/ext/filters/client_channel/lb_policy/rls/rls.cc#L2380

There's also a use-case where we need to do a custom check on the value of a string field:

https://github.com/grpc/grpc/blob/28db143b98682b06aa5c58059ffe53b42f28b395/src/core/ext/filters/client_channel/lb_policy/rls/rls.cc#L2349

In those two particular cases, it probably wouldn't be terrible to just do that validation after the parsing is done.  But I could imagine cases where we need to do that validation for an object that is nested deep within the JSON hierarchy in many different places, where it might make sense to do that validation after parsing each instance of the object.  Is there a way to add a post-parsing hook here that can look at the fields in the resulting struct and optionally add errors if there's a validation failure?

Reply:
Added an optional JsonPostLoad function that can do things after the object is loaded but before it's added to a parent